x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/IndicesAccessControl.java

```diff
@@ -64,12 +72,29 @@ public boolean isGranted() {
         return granted;
     }
 
-    public Collection<?> getDeniedIndices() {
-        return this.indexPermissions.entrySet()
-            .stream()
-            .filter(e -> e.getValue().granted == false)
-            .map(Map.Entry::getKey)
-            .collect(Collectors.toUnmodifiableSet());
+    /**
+     * @return A collection on index names for which the current request was denied access, divide in regular and
+     * {@link #isRestrictedIndex(String) restricted} indices.
+     */
+    public DeniedIndices getDeniedIndices() {
+        final List<String> regular = new ArrayList<>();
+        final List<String> restricted = new ArrayList<>();
+        this.indexPermissions.entrySet().stream().filter(e -> e.getValue().granted == false).map(Map.Entry::getKey).forEach(name -> {
+            if (isRestrictedIndex(name)) {
+                restricted.add(name);
+            } else {
+                regular.add(name);
+            }
+        });
+        return new DeniedIndices(Set.copyOf(regular), Set.copyOf(restricted));
```

Comment:
Why allocate `ArrayList`s if we're going to copy to a set anyway? We could either:
1. Allocate `HashSet`s instead and just wrap them with `Collections.unmodifiableSet`, which saves a copy of the collection, or
2. Use `Collectors.groupingBy(this::isRestrictedIndex, Collectors.toUnmodifiableSet())` instead of the `forEach`.

Reply:
> Why allocate ArrayLists if we're going to copy to a set anyway? 

Because I can never remember where `Collectors.groupingBy` is, and assume I must have imagined its existence. In Scala the `groupedBy` method is much more obvious, and when I can't find the Java equivalent one I assume it's the Scala part of my brain seeking revenge for the abuse it has suffered.

Thanks for the pointer, maybe I'll remember it this time.

Reply:
FWIW, I'm not sure than an `ArrayList` + `Set.copyOf` is actually any less efficient than `Collectors.toUnmodifiedSet`

Unless, I've misread something the former will:
1. Collect items in a List backed by an array, which means there's a potential resize cost, but it's otherwise pretty efficient
2. Turn the array list into a `HashSet` (see `Set.copyOf`) which ends up doing a item-by-item `add` which will incur a hash / duplicate check on each add 
3. Turn the `HashSet` into an array (see `Set.copyOf`)
4. Wrap a `ImmutableCollections.SetN` around the array (assuming >2 elements)

The latter will:
1. Collect items in a `HashSet`, which which ends up doing a item-by-item `add` which will incur a hash / duplicate check on each add 
2. Turn that `HashSet` into another `HashSet` (see `Set.copyOf`) which ends up doing a item-by-item `add` which will incur another hash / duplicate check on each add 
3. Turn the `HashSet` into an array (see `Set.copyOf`)
4. Wrap a `ImmutableCollections.SetN` around the array (assuming >2 elements)

Ultimately, it's only step 1 that's any different, and the ArrayList may end up being lower cost.

Reply:
Ah, you're right! I incorrectly assumed that `Collectors.toUnmodifiableSet` would just be `toSet` with its output wrapped in a `Collections.UnmodifiableSet`, which does not copy the collection and is instead just delegates to the original instance. So either my first suggestion or using `Collectors.toSet` and manually wrapping it with an `UnmodifiableSet` would save the copy of the set I think, but this is pretty nitpicky anyway so I'm not going to press too hard for one of those (especially as, with very small collections like I'd expect here, a `SetN` very well might be more efficient than a `HashSet` anyway)